components/service/glean/src/main/java/mozilla/components/service/glean/storages/EventsStorageEngine.kt

```diff
@@ -70,6 +71,9 @@ internal object EventsStorageEngine : StorageEngine {
             for (storeName in stores) {
                 val storeData = eventStores.getOrPut(storeName) { mutableListOf() }
                 storeData.add(event.copy())
+                if (storeData.size >= Glean.configuration.maxEvents) {
+                    Glean.sendPing(storeName, "events")
```

Comment:
Would this guarantee that only `maxEvents` events make it into the ping?

Reply:
In practice, yes.  Using `>=` rather than `==` is just insurance against another future code path adding events without checking and sending, but hopefully that would be caught be testing.

Reply:
Thinking this through again.  Given that ping building is async, there is a theoretical condition where:

1) The 500th event is recorded, triggering an eventual ping build
2) A 501st event is recorded
3) The ping is built

This would lead to a ping with 501 events, which I think is fine and preferable to data loss and preferable to blocking on ping building.

But this does lead me to feel that the `>=` should change to `==` to prevent two pings from being sent in this scenario.

Reply:
Good point. I think we always prefer 501 to data loss and blocking. Maybe we could check how focus and telemetry-android deal with this?

Reply:
So, looks like telemetry-android has a [similar behaviour](https://github.com/mozilla-mobile/android-components/blob/bc0f57b553a41f1e93213937118f6c84a442803e/components/service/telemetry/src/main/java/org/mozilla/telemetry/Telemetry.java#L112-L113). Moreover, by looking at [FFTV](https://sql.telemetry.mozilla.org/queries/60515/source#156227), it seems that we don't have that many clients hitting the limit... so is an edge case of a limit case? Which means we should be pretty much ok?

Reply:
Yeah.  I think particularly on mobile where we are sending/clearing the event queue every time we go to background, we're even more likely to not fill up the queue than on desktop.  So I'm pretty confident we are good here.